### PR TITLE
Enable support for Istanbul

### LINKED ISSUE #7310: Support for Istanbul
For an accurate list of changes see [EIP-1679](https://eips.ethereum.org/EIPS/eip-1679).

Valid as of today:
- EIP-152: Add Blake2 compression function F precompile
- EIP-1108: Reduce alt_bn128 precompile gas costs
- EIP-1344: Add ChainID opcode
- EIP-1884: Repricing for trie-size-dependent opcodes
- EIP-2028: Calldata gas cost reduction
- EIP-2200: Rebalance net-metered SSTORE gas cost with consideration of SLOAD gas cost change

Changes needed in Solidity:
- [x] Add Istanbul EVMVersion [#7329]
- [x] Add `chainid` opcode [#7327]
- [x] Add `selfbalance` opcode [#7327]
- [x] Make use of `selfbalance` opcode.
- [x] Reprice call data (EIP-2028), `sstore` (EIP-2200), `sload`/`balance`/`extcodehash` (EIP-1884), bn128 precompiles (in tests) (EIP-1108) [#7740]
- [x] Update evmc/evmone [#7655]

Please note that evmone currently doesn't implement any of the Istanbul changes and the breaking EVMC7 release is required for Istanbul support.

### PR DESCRIPTION
**Pull request by axic**
Closes #7310.

Depends on #7655.

### CODE REVIEW

**Comment by axic** [comment]
Updated test expectations in GasCosts.

File: test/EVMHost.h [hunk 1]
```diff
@@ -180,7 +180,8 @@ class EVMHost: public evmc::Host
 	static evmc::result resultWithGas(evmc_message const& _message, bytes const& _data) noexcept;
 
 	evmc::VM* m_vm = nullptr;
-	evmc_revision m_evmVersion;
+	langutil::EVMVersion m_evmVersion;
+	evmc_revision m_evmRevision;
```
Comment (Marenz) [inline test/EVMHost.h:186]: Could you add two comments explaining the difference between revision and version?

File: scripts/tests.sh [hunk 2]
```diff
@@ -81,7 +81,7 @@ EVM_VERSIONS="homestead byzantium"
 
 if [ -z "$CI" ]
 then
-    EVM_VERSIONS+=" constantinople petersburg"
+    EVM_VERSIONS+=" constantinople petersburg istanbul"
 fi
 
 # And then run the Solidity unit-tests in the matrix combination of optimizer / no optimizer
```
Comment (Marenz) [inline scripts/tests.sh:87]: Is the comment below this line still correct?

**Review by Marenz — CHANGES_REQUESTED** [review]

File: test/EVMHost.cpp [hunk 3]
```diff
@@ -64,7 +64,7 @@ evmc::VM* EVMHost::getVM(string const& _path)
 }
 
 EVMHost::EVMHost(langutil::EVMVersion _evmVersion, evmc::VM* _vm):
-	m_vm(_vm)
+	m_vm(_vm), m_evmVersion(_evmVersion)
```
Comment (leonardoalt) [inline test/EVMHost.cpp]:
```suggestion
	m_vm(_vm),
	m_evmVersion(_evmVersion)
```

File: test/externalTests/common.sh [hunk 4]
```diff
@@ -207,7 +207,7 @@ function run_test
     for optimize in "${optimizer_settings[@]}"
     do
         clean
-        force_solc_settings "$CONFIG" "$optimize" "petersburg"
+        force_solc_settings "$CONFIG" "$optimize" "istanbul"
```
Comment (erak) [inline test/externalTests/common.sh:210]: This will probably break some tests, but what could one do :wink:

**Review by erak — APPROVED** [review]

File: test/externalTests/common.sh (hunk 4 above)
Comment (axic) [reply test/externalTests/common.sh:210]: I'm unsure about the changes in the test scripts.

File: test/libsolidity/GasCosts.cpp [hunk 5]
```diff
@@ -104,9 +104,19 @@ BOOST_AUTO_TEST_CASE(string_storage)
 	else if (Options::get().useABIEncoderV2)
 	{
 		if (Options::get().optimizeYul)
-			CHECK_DEPLOY_GAS(151455, 127653, evmVersion);
+		{
+			if (evmVersion < EVMVersion::istanbul())
+				CHECK_DEPLOY_GAS(151455, 127653, evmVersion);
+			else
+				CHECK_DEPLOY_GAS(123, 113821, evmVersion);
```
Comment (chriseth) [inline test/libsolidity/GasCosts.cpp]: The `123` here is just not hit by the tests, right?

File: test/libsolidity/GasCosts.cpp [hunk 6]
```diff
@@ -104,9 +104,19 @@ BOOST_AUTO_TEST_CASE(string_storage)
 	else if (Options::get().useABIEncoderV2)
 	{
 		if (Options::get().optimizeYul)
-			CHECK_DEPLOY_GAS(151455, 127653, evmVersion);
+		{
+			if (evmVersion < EVMVersion::istanbul())
```
Comment (chriseth) [inline test/libsolidity/GasCosts.cpp:109]: It might have been better to directly take the data gas cost change into account and keep the constants.

**Review by chriseth — APPROVED** [review]

File: test/libsolidity/GasCosts.cpp (hunk 5 above)
Comment (axic) [reply test/libsolidity/GasCosts.cpp]: Yes, marked them as 0 in a commit.

**Pull request merged**